test/tests/functional/pbs_node_rampdown.py

```diff
@@ -6362,3 +6415,774 @@ def test_release_nodes_shared_server_restart_immed(self):
+        attr1['type'] = 'size'
+        attr1['flag'] = 'nh'
+        r1 = 'foo_i'
+        self.server.manager(
+            MGR_CMD_CREATE, RSC, attr1, id=r1, runas=ROOT_USER, logerr=False)
+
+        self.logger.info("test_epilogue")
+        hook_body = """
+import pbs
+e=pbs.event()
+pbs.logmsg(pbs.LOG_DEBUG, "executed epilogue hook")
+if e.job.in_ms_mom():
+    e.job.resources_used["vmem"] = pbs.size("9gb")
+    e.job.resources_used["foo_i"] = pbs.size(999)
+    e.job.resources_used["foo_f"] = 0.09
+else:
+    e.job.resources_used["vmem"] = pbs.size("10gb")
+    e.job.resources_used["foo_i"] = pbs.size(1000)
+    e.job.resources_used["foo_f"] = 0.10
+"""
+
+        hook_name = "epi"
+        a = {'event': "execjob_epilogue", 'enabled': 'True'}
+        rv = self.server.create_import_hook(
+            hook_name,
+            a,
+            hook_body,
+            overwrite=True)
+        self.assertTrue(rv)
+
+        jid = self.create_and_submit_job('job1_5')
+        self.server.expect(JOB, {'job_state': "R"}, id=jid)
+
+        # Release hostC
+        cmd = [self.pbs_release_nodes_cmd, '-j', jid, self.n7]
+        ret = self.server.du.run_cmd(self.server.hostname, cmd)
+        self.assertEqual(ret['rc'], 0)
+
+        self.momC.log_match("executed epilogue hook", max_attempts=10)
+        self.momC.log_match("DELETE_JOB2 received", max_attempts=10)
+
+        self.server.delete(jid, wait=True)
+
+        self.server.expect(JOB, {'job_state': 'F',
+                                 "resources_used.foo_i": "3kb",
+                                 "resources_used.foo_f": '0.29',
+                                 "resources_used.vmem": '29gb'}, id=jid)
+
+    @timeout(500)
+    def test_release_reservations(self):
+        """
+        Release nodes from a reservation wil throw error. However
+        jobs inside reservation queue works as expected.
+        """
+
+        # Create a reservation on multiple nodes
+        start = int(time.time()) + 60
+        a = {'Resource_List.select': self.job1_select,
+             'Resource_List.place': 'scatter',
+             'reserve_start': start}
+        r = Reservation(TEST_USER, a)
+        rid = self.server.submit(r)
+        rid = rid.split('.')[0]
+
+        self.server.expect(RESV,
+                           {'reserve_state': (MATCH_RE, "RESV_CONFIRMED|2")},
+                           id=rid)
+
+        # Release a vnode from reservation. It will throw error.
+        cmd = [self.pbs_release_nodes_cmd, '-j', rid, self.n5]
+        r = self.server.du.run_cmd(self.server.hostname, cmd)
+        self.assertNotEqual(r['rc'], 0)
+
+        # Submit a job inside reservations and release vnode
+        a = {'queue': rid,
+             'Resource_List.select': self.job1_select}
+        j = Job(TEST_USER, a)
+        jid = self.server.submit(j)
+
+        # Wait for the job to start
+        self.server.expect(JOB, {'job_state': 'R'},
+                           offset=60, id=jid)
```

Comment:
Can we shorten this to 30 (and reserve_start) or will that cause problems?

Reply:
let me try. Reason I kept 60 as I have noticed pbs_rsub will round off the start time and not actually now+30 causing race conditions in the test.